### PR TITLE
Added filter by type

### PR DESCRIPTION
**Pull request by datanel**
This PR uses https://github.com/datanel/queryst?branch=decode-params merged (https://github.com/rustless/queryst/pull/8) but no new release was  created

### TODO
- [x] integration tests

### CODE REVIEW

File: libs/bragi/src/query.rs [hunk 1]
```diff
@@ -199,29 +199,76 @@ fn build_query(q: &str,
...
-            result.push(dataset);
+
+    let mut result: Vec<String> = vec![];
+    if let Some(ref types) = *types {
+        let mut indexes = vec![];
+        for a_type in types {
+            if *a_type != "public_transport:stop_area" {
+                indexes.push(get_indexes_by_type(a_type));
+            }
+        }
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]:
```rust
for type_ in types.iter().filter(|t| t != "public_transport:stop_area") {
    result_push(get_indexes_by_type(t));
}
```

File: libs/bragi/src/query.rs [hunk 2]
```diff
@@ -199,29 +199,76 @@ fn build_query(q: &str,
...
+        if pt_dataset_index.is_some() && types.contains(&"public_transport:stop_area") {
+            indexes.push(pt_dataset_index.clone().unwrap());
+        }
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]:
```rust
match pt_dataset_index {
    Some(ref index) if types.contains(&"public_transport:stop_area") => result_push(index),
    _ => (),
}
```
Or something else avoiding the `unwrap()`

File: libs/bragi/src/query.rs [hunk 3]
```diff
@@ -199,29 +199,76 @@ fn build_query(q: &str,
...
     }
-    let mut result: Vec<&str> = vec![];
-    if is_existing_index(client, &"munin_geo_data".to_string()) {
-        result.push("munin_geo_data");
-    }
-    if let Some(ref dataset) = *pt_dataset_index {
-        if is_existing_index(client, dataset) {
-            result.push(dataset);
+
+    let mut result: Vec<String> = vec![];
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]:
```rust
let result_push = |i: &str| {
    if is_existing_index(i) {
        result.push(i.into());
    }
};
```

File: libs/bragi/src/query.rs [hunk 4]
```diff
@@ -199,29 +199,76 @@ fn build_query(q: &str,
...
+            if *a_type != "public_transport:stop_area" {
+                indexes.push(get_indexes_by_type(a_type));
+            }
+        }
+        if pt_dataset_index.is_some() && types.contains(&"public_transport:stop_area") {
+            indexes.push(pt_dataset_index.clone().unwrap());
+        }
+
+        if indexes.is_empty() {
+            indexes.push("munin_geo_data".to_string());
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: and so on

File: libs/bragi/src/query.rs [hunk 5]
```diff
@@ -199,29 +199,76 @@ fn build_query(q: &str,
...
+        }
+
+        for index in indexes {
+            if is_existing_index(index.as_str()) {
+                result.push(index);
+            }
+        }
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: and delete that.

File: libs/bragi/src/query.rs [hunk 6]
```diff
@@ -199,29 +199,76 @@ fn build_query(q: &str,
...
+    } else {
+        if is_existing_index("munin_geo_data") {
+            result.push("munin_geo_data".to_string());
+        }
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: and so on

File: libs/bragi/src/query.rs [hunk 7]
```diff
@@ -199,29 +199,76 @@ fn build_query(q: &str,
...
+        if let Some(ref dataset) = *pt_dataset_index {
+            if is_existing_index(dataset.as_str()) {
+                result.push(dataset.clone());
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: and so on

File: libs/bragi/src/query.rs [hunk 8]
```diff
@@ -199,29 +199,76 @@ fn build_query(q: &str,
...
     }
+
     result
 }
 
+fn make_indexes<'a>(all_data: bool,
+                    pt_dataset_index: &'a Option<String>,
+                    types: &Option<Vec<&str>>,
+                    client: &mut rs_es::Client)
+                    -> Vec<String> {
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: lifetime `'a` is now useless

File: libs/bragi/src/query.rs [hunk 9]
```diff
@@ -199,29 +199,76 @@ fn build_query(q: &str,
...
+        "public_transport:stop_area" => "stop",
+        "city" => "admin",
+        "house" => "addr",
+        _ => a_type,
+    };
+
+    format!("munin_{}", doc_type)
+}
+
+fn make_indexes_impl<'a, F: FnMut(&str) -> bool>(all_data: bool,
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: lifetime is now useless

File: tests/bragi_filter_types_test.rs [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    assert_eq!(count, 0);
+}
+
+fn type_stop_area_no_dataset_test(bragi: &BragiHandler) {
+    // with this query we should not find any stops
+    let response = bragi.get("/autocomplete?q=14 juillet&type[]=public_transport:stop_area");
+    let types = get_types(&response);
+    let count = count_types(&types, "public_transport:stop_area");
+    assert_eq!(count, 0);
+}
```
Comment (TeXitoi) [inline tests/bragi_filter_types_test.rs]: Please a test when we find a stop_area.

File: libs/bragi/src/api.rs [hunk 11]
```diff
@@ -242,14 +248,20 @@ impl ApiEndPoint {
                             lat: lat.unwrap(),
                         })
                     });
+
+                    let types = params.find("type")
+                        .and_then(|val| val.as_array())
+                        .map(|val| val.iter().map(|val| val.as_str().unwrap()).collect());
```
Comment (datanel) [inline libs/bragi/src/api.rs]: one for api.post and another for api.get

File: libs/bragi/Cargo.toml [hunk 12]
```diff
@@ -22,5 +22,8 @@ serde_derive = "0.8"
 geojson = { version="*", features=["with-serde"]}
 geo = "^0.3"
 
+# remove this line once is rustless > 0.10.0 is released
```
Comment (TeXitoi) [inline libs/bragi/Cargo.toml]: queryst > 1.0.2

File: libs/bragi/src/api.rs [hunk 13]
```diff
@@ -56,6 +56,28 @@ fn render<T>(mut client: rustless::Client,
...
+        match type_ {
+            "city" => Some(Type::City),
+            "house" => Some(Type::House),
+            "poi" => Some(Type::Poi),
+            "public_transport:stop_area" => Some(Type::StopArea),
+            "street" => Some(Type::Street),
+            _ => None,
+        }
+    }
+}
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]: minor: you can `impl std::str::FromStr`

**Pull request merged**